In [1]:
# ==========================================================
# IMPORTAÇÃO DAS BIBLIOTECAS
# ==========================================================

import pandas as pd

# Divide os dados em treino e teste
from sklearn.model_selection import train_test_split

# Faz a normalização dos dados
from sklearn.preprocessing import MinMaxScaler

# Algoritmo KNN pronto da biblioteca scikit-learn
from sklearn.neighbors import KNeighborsClassifier

# Métricas de avaliação
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [2]:
# ==========================================================
# LEITURA DO DATASET
# ==========================================================

df = pd.read_csv("https://drive.google.com/uc?export=download&id=1wF7amDwgNKIxa9Lc5Dm4OXkiDy_N_CYu")

print("Formato do dataset:", df.shape)
print(df.head())

Formato do dataset: (1998, 20)
   battery_power  bluetooth  clock_speed  dual_sim  front_camera  four_g  \
0            667          1          0.6         0             6       1   
1           1162          0          2.0         0             0       1   
2           1712          1          0.9         1             2       0   
3           1696          0          1.7         0             1       0   
4            592          0          0.8         1             1       0   

   storage  mobile_weight  n_cores  primary_camera  px_height  px_width   ram  \
0        3            185        6              13         13      1463  2855   
1       56            114        7               1         56       997  1652   
2       23            155        7              19        155       663  1422   
3       13             97        3               2        179       772   523   
4       52            103        7              10        332       970  1507   

   sc_height  sc_width  t

In [4]:
# ==========================================================
# ESCOLHA DOS ATRIBUTOS
# ==========================================================

# Mesmas colunas usadas no KNN manual
colunas_usadas = [
"battery_power",
    "clock_speed",
    "n_cores",
    "ram"
]

# X = atributos usados para prever
X = df[colunas_usadas]

# y = variável alvo
y = df["price_range"]

In [5]:
# ==========================================================
# DIVISÃO ENTRE TREINO E TESTE
# ==========================================================

# 80% treino e 20% teste
# random_state mantém o mesmo sorteio sempre que rodar
# stratify=y mantém a proporção das classes em treino e teste
X_treino, X_teste, y_treino, y_teste = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("Quantidade treino:", len(X_treino))
print("Quantidade teste:", len(X_teste))

Quantidade treino: 1598
Quantidade teste: 400


In [6]:
# ==========================================================
# NORMALIZAÇÃO DOS DADOS
# ==========================================================

# Cria o normalizador Min-Max
normalizador = MinMaxScaler()

# Aprende os mínimos e máximos usando apenas os dados de treino
X_treino_normalizado = normalizador.fit_transform(X_treino)

# Aplica a mesma normalização nos dados de teste
X_teste_normalizado = normalizador.transform(X_teste)

print("Dados normalizados com sucesso!")

Dados normalizados com sucesso!


In [43]:
# ==========================================================
# CRIAÇÃO DO MODELO KNN
# ==========================================================

# Define o valor de K
k = 100

# Cria o modelo KNN
modelo_knn = KNeighborsClassifier(n_neighbors=k)

print("Valor escolhido para K:", k)

Valor escolhido para K: 100


In [44]:
# ==========================================================
# TREINAMENTO DO MODELO
# ==========================================================

# O modelo aprende usando os dados de treino
modelo_knn.fit(X_treino_normalizado, y_treino)

KNeighborsClassifier(n_neighbors=100)

In [45]:
# ==========================================================
# PREVISÕES
# ==========================================================

# O modelo faz previsões nos dados de teste
previsoes = modelo_knn.predict(X_teste_normalizado)

In [46]:
# ==========================================================
# AVALIAÇÃO DO MODELO
# ==========================================================

# Calcula a acurácia
acuracia = accuracy_score(y_teste, previsoes)

print("\n================ RESULTADOS ================")
print("Total de testes:", len(y_teste))
print("Acurácia:", acuracia)
print("Acurácia em porcentagem:", round(acuracia * 100, 2), "%")


================ RESULTADOS ================
Total de testes: 400
Acurácia: 0.7675
Acurácia em porcentagem: 76.75 %


In [47]:
# ==========================================================
# MATRIZ DE CONFUSÃO
# ==========================================================

print("\nMatriz de confusão:")
print(confusion_matrix(y_teste, previsoes))


Matriz de confusão:
[[83 17  0  0]
 [ 9 77 14  0]
 [ 0 20 63 17]
 [ 0  0 16 84]]


In [48]:
# ==========================================================
# RELATÓRIO DE CLASSIFICAÇÃO
# ==========================================================

print("\nRelatório de classificação:")
print(classification_report(y_teste, previsoes))


Relatório de classificação:
              precision    recall  f1-score   support

           0       0.90      0.83      0.86       100
           1       0.68      0.77      0.72       100
           2       0.68      0.63      0.65       100
           3       0.83      0.84      0.84       100

    accuracy                           0.77       400
   macro avg       0.77      0.77      0.77       400
weighted avg       0.77      0.77      0.77       400



### Relatórios de classificação

```text

K=1
              precision    recall  f1-score   support

           0       0.82      0.79      0.81       100
           1       0.58      0.65      0.61       100
           2       0.64      0.58      0.61       100
           3       0.83      0.84      0.84       100

    accuracy                           0.71       400
   macro avg       0.72      0.71      0.72       400
weighted avg       0.72      0.71      0.72       400

K=5
Relatório de classificação:
              precision    recall  f1-score   support

           0       0.84      0.81      0.82       100
           1       0.65      0.70      0.68       100
           2       0.71      0.67      0.69       100
           3       0.85      0.86      0.86       100

    accuracy                           0.76       400
   macro avg       0.76      0.76      0.76       400
weighted avg       0.76      0.76      0.76       400

K=15
Relatório de classificação:
              precision    recall  f1-score   support

           0       0.86      0.86      0.86       100
           1       0.69      0.74      0.71       100
           2       0.71      0.67      0.69       100
           3       0.86      0.85      0.85       100

    accuracy                           0.78       400
   macro avg       0.78      0.78      0.78       400
weighted avg       0.78      0.78      0.78       400

K=25
Relatório de classificação:
              precision    recall  f1-score   support

           0       0.88      0.87      0.87       100
           1       0.69      0.75      0.72       100
           2       0.72      0.66      0.69       100
           3       0.87      0.87      0.87       100

    accuracy                           0.79       400
   macro avg       0.79      0.79      0.79       400
weighted avg       0.79      0.79      0.79       400

K=50
Relatório de classificação:
              precision    recall  f1-score   support

           0       0.91      0.86      0.89       100
           1       0.71      0.79      0.75       100
           2       0.71      0.67      0.69       100
           3       0.86      0.86      0.86       100

    accuracy                           0.80       400
   macro avg       0.80      0.79      0.80       400
weighted avg       0.80      0.80      0.80       400

K=100
Relatório de classificação:
              precision    recall  f1-score   support

           0       0.90      0.83      0.86       100
           1       0.68      0.77      0.72       100
           2       0.68      0.63      0.65       100
           3       0.83      0.84      0.84       100

    accuracy                           0.77       400
   macro avg       0.77      0.77      0.77       400
weighted avg       0.77      0.77      0.77       400

K=150
Relatório de classificação:
              precision    recall  f1-score   support

           0       0.89      0.80      0.84       100
           1       0.64      0.72      0.68       100
           2       0.68      0.64      0.66       100
           3       0.85      0.88      0.86       100

    accuracy                           0.76       400
   macro avg       0.76      0.76      0.76       400
weighted avg       0.76      0.76      0.76       400